helix-term/src/ui/editor.rs

```diff
@@ -264,6 +272,121 @@ impl EditorView {
+            .rev()
+            .find(|&i| !is_word(text.char(i)))
+            .map(|i| i + 1)
+            .unwrap_or(0);
+        let end = (cursor_char..text.len_chars())
+            .find(|&i| !is_word(text.char(i)))
+            .unwrap_or_else(|| text.len_chars());
+
+        Some(text.slice(start..end).to_string())
+    }
+
+    /// Calculates the ranges of the word under the cursor and returns the result
+    fn calculate_cursor_word(
+        doc: &Document,
+        view: &View,
+        viewport: Rect,
+        scope_index: usize,
+    ) -> Vec<(usize, std::ops::Range<usize>)> {
+        let text = doc.text().slice(..);
+        let mut result = Vec::new();
+
+        let cursor_word = match Self::cursor_word(doc, view) {
+            Some(cw) => cw,
+            None => return result,
+        };
+
+        let is_word_char = |c: Option<char>| -> bool {
+            if let Some(c) = c {
+                c.is_ascii_alphanumeric() || c == '-' || c == '_'
+            } else {
+                false
+            }
+        };
+
+        let row = text.char_to_line(view.offset.anchor.min(text.len_chars()));
+
+        let line_range = {
+            // Calculate the lines in view
+            let last_line = text.len_lines().saturating_sub(1);
+            let last_visible_line = (row + viewport.height as usize).min(last_line);
+            let first_visible_line = row;
+
+            first_visible_line..last_visible_line
+        };
+
+        for line in line_range {
+            let current_line = text.line(line);
+            let current_line = match current_line.as_str() {
+                Some(l) => l,
+                None => continue,
+            };
+
+            let line_offset = text.line_to_char(line);
+
+            result.extend(
+                current_line
+                    .match_indices(&cursor_word)
+                    .map(|(i, _)| i)
+                    .filter(|i| {
+                        if *i != 0 {
+                            !is_word_char(current_line.chars().nth(i - 1))
+                        } else {
+                            true
+                        }
+                    })
+                    .filter(|i| !is_word_char(current_line.chars().nth(i + cursor_word.len())))
+                    .map(|i| line_offset + i)
+                    .map(|start| (scope_index, { start..start + cursor_word.len() })),
+            );
+        }
+
+        result
+    }
+
+    /// Apply the decoration for the word to be highlighted
+    pub fn collect_cursor_word_highlights(
+        doc: &Document,
+        editor: &Editor,
+        view: &View,
+        viewport: Rect,
+        theme: &Theme,
+    ) -> Option<Vec<(usize, std::ops::Range<usize>)>> {
+        let scope_index = theme.find_scope_index("ui.wordmatch")?;
+        let mut result: Vec<(usize, std::ops::Range<usize>)> = Vec::new();
+        let is_lsp_ready = doc.language_server().is_some() && !editor.cursor_highlights.is_empty();
```

Comment:
I think there should be some specical handling to indicate when the LSP is ready. It's perfectly valid to for the LSP to return an empty range and if that is really the LSPs newest response than we we should do respect that. An option would probably work well

Reply:
I am not sure if my updated check works as expected, but i think it's as close to figuring out whether lsp positions should be used or not.